From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheKit <nekit1000@gmail.com>
Date: Mon, 20 Mar 2023 18:30:57 +0000
Subject: [PATCH] (halium) libprocessgroup: fix mounting cpu/cpuacct on systemd
 hosts

systemd with cgroup v1 mounts cpu and cpuacct controllers together.
As result, libprocessgroup's attempt to mount them independently will fail.
If this happens, retry and mount cpu and cpuacct combined instead.

Change-Id: I7893b0a809dffec49c15f36621363552d2e08667
---
 libprocessgroup/setup/cgroup_map_write.cpp | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/libprocessgroup/setup/cgroup_map_write.cpp b/libprocessgroup/setup/cgroup_map_write.cpp
index 899b4e9524..a6f43f9a1b 100644
--- a/libprocessgroup/setup/cgroup_map_write.cpp
+++ b/libprocessgroup/setup/cgroup_map_write.cpp
@@ -154,8 +154,19 @@ static bool MountV2CgroupController(const CgroupDescriptor& descriptor) {
         LOG(INFO) << "Mounting memcg with memory_recursiveprot failed. Retrying without.";
         if (mount("none", controller->path(), "cgroup2", MS_NODEV | MS_NOEXEC | MS_NOSUID,
                   nullptr) < 0) {
-            PLOG(ERROR) << "Failed to mount cgroup v2";
-            return IsOptionalController(controller);
+            PLOG(WARNING) << "Failed to mount cgroup v2";
+
+            // Halium: systemd with cgroup v1 mounts cpu and cpuacct controllers together.
+            // As result, libprocessgroup's attempt to mount them independently will fail.
+            // If this happens, retry and mount cpu and cpuacct together.
+            if (!strcmp(controller->name(), "cpu") || !strcmp(controller->name(), "cpuacct")) {
+                if (mount("none", controller->path(), "cgroup", MS_NODEV | MS_NOEXEC | MS_NOSUID,
+                          "cpu,cpuacct") < 0) {
+
+                    PLOG(ERROR) << "Failed to mount cgroup v1 with Halium hack";
+                    return IsOptionalController(controller);
+                }
+            }
         }
     }
 
